### ISSUE TITLE
VB WellKnownFixAllProviders.BatchFixer runs out of Memory

### ISSUE THREAD

**Issue by paul1956** [dotnet/roslyn#22539]
**Version Used**: 
VS 2017 15.3.5 All Roslyn Versions
**Steps to Reproduce**:

1. Use Any CodeFix that can fix a lot of issues, for example add AsClause to all of Roslyn
2. You will run out of memory after a few 100 fixes (in most cases you can do a document worth but project never succeeds).

**Expected Behavior**:
I would be nice if the WellKnownFixAllProviders.BatchFixer  stopped before it ran out of memory even if It didn't fix every issue in a Document/Project/Solution.
**Actual Behavior**:
When applying a BatchFix that effects 100's of files you get a System,OutOfMemoryException and the CodeFix will exit. The current release of Roslyn 2.x is much better than 1.x which failed before a single document with a few 100 fixes.

**Comment by AdamSpeight2008** [comment]
Sounds like it is creating a new document for each fix and not reusing an temporary document for each successive fix. Akin to a aggregate function in Linq.
```c#
Document FixAll(Document curr, Func<Document, Document> fixer )
{
  return Enumerable.Aggregate( curr, (newDoc,fixer)=> fixer(UpdatedDocument));
}
```

**Comment by paul1956** [comment]
Something changed in 15.4/2.4 I am able to run 1,000's of fixes without running out of memory of course it takes days but its progress.

**Comment by CyrusNajmabadi** [comment]
I rewrote the stock batch fixer to to be better.   There is a general problem that our entire system is based on creating a new snapshot and then applying the snapshot all at once (i.e. better to make no changes than to make a partial change that messes up the user's solution).  If that new solution snapshot is sufficiently different from the original solution, it is conceivable that it would take up too much memory and could die.

I would recommend that if you are making a huge change, that you *not* use the stock batch fixer.  It's just not great at this sort of thing.  It's just a acceptable for many (but not all) use cases.

For something that needs to make huge solution-level changes, i recommend taking an approach similar to what we have in https://github.com/dotnet/roslyn/blob/master/src/Workspaces/Core/Portable/CodeFixes/SyntaxEditorBasedCodeFixProvider.FixAllProvider.cs 

Here, instead of producing a set of text changes and then attempting to merge all the disparate edits,  the fixer just makes all the tree changes individually to a SyntaxEditor which can then apply all the changes bottom-up in a single pass.

For performance, we still process all documents for a project in parallel, that way we can get a linear speedup with the number of cores you have.

**Comment by CyrusNajmabadi** [comment]
> Sounds like it is creating a new document for each fix and not reusing an temporary document for each successive fix.

Yes.  It has to work in that manner.  The fixes that are aggregates are all created against the original document.  If you tried to apply them to a changed document then all sorts of bad things could happen.

--

Also, to be clear, there's no way to even pass in 'a temporary document' to the fixers.  The fixers completely control what they do.  Many are just a ```Func<CancellationToken, Task<Document>>```.  Meaning that all they do when invoked is just return the Document they want to change the *original* document to.  All they are fed is a CancellationToken.  

--

In order to get the benefits of 'a temporary document'.  The fixer needs to supply it's own FixAllProvider.  That provider will get all the diagnostics, and can decide how to apply them.   That's how the SyntaxEditorBasedCodeFixProvider works.  It gets all the diagnostics, creates a single editor, makes all the changes to that editor, and then applies it all at once.

**Comment by paul1956** [comment]
@CyrusNajmabadi  Most of the code called by https://github.com/dotnet/roslyn/blob/master/src/Workspaces/Core/Portable/CodeFixes/SyntaxEditorBasedCodeFixProvider.FixAllProvider.cs is marked Friend and not callable by my CodeFix unless I port everything from Roslyn Source, I find this true with much of the Sample code as well.

**Comment by CyrusNajmabadi** [comment]
Right, that's why i said "i recommend taking an approach *similar* to what we have in" :)

That type hasn't been vetted as being appropriate for 3rd party consumption.  it's been written to work against the specific patterns and approaches we've taken with many of our fixers (i.e. being SyntaxEditor dependendent).  As such, it's not necessarily reusable for other situations.

My point was, if you create a fixer that does something like "add AsClause to all of Roslyn", then i would recommend not using the stock batch fixer for this.  It's just not going to scale as well as something custom built to go and actually make this change.  The stock batch fixer is for simpler changes, and for providing a cheap and easy way to provide 'fix all'.  But if you want a very scalable solution, it isn't necessarily the best choice.

**Comment by paul1956** [comment]
Following the "example" of SyntaxEditorBasedCodeFix sends one does a serious rabbit hole. It depends on a slew of other internal code which I also question being internal, why would ArrayBuilder or SpecializedCollection be internal. I would like to see it made public and let the user beware. Using it for the wrong situations isn't any worse than using the current WellKnownFixAllProviders.BatchFixer for the wrong situations.

**Comment by CyrusNajmabadi** [comment]
> Following the "example" of SyntaxEditorBasedCodeFix sends one does a serious rabbit hole. It depends on a slew of other internal code which I also question being internal, 

I didn't say "take the code as is and use it", i said that you should look to it as an example of how you can write a more domain specific batch fixer that can do better in certain circumstances.  The general design is simply: "Get your diagnostics, create a syntax editor.  Process each diagnostic, making whatever changes you want to that editor.  Finally return the total edited tree."   The actual code you use to do that can be your own and can be whipped up rather simply.

> why would ArrayBuilder extensions or SpecializedCollection be internal. 

Because public APIs incur a maintenance cost.  You can't change them as they could now break 3rd party consumers.  We can and do change our internal code and we don't want to be locked into these internal implementation details that are not part of the core value that Roslyn is providing.

**Comment by paul1956** [comment]
@CyrusNajmabadi  the issue is the examples and SyntaxEditorBasedCodeFix use the internal API's so to use them as a starting point I need to clone the internal API's. Also using SyntaxEditor as a starting point is not something that can be "whipped up rather simply" without internal knowledge even Bing can't find documentation on usage. All that said, with the fixes in 2.4 I can process 1,000's of code fixes with any noticeable consumption of memory, it could also be that I now ignore Generated code so I don't try to process the huge generated files. Of course the Generated Code detection is also internal. Since I am trying to fix standard VB errors like ERR_StrictDisallowImplicitObject and it does not ignore Generated Code my code fix provider needs to do it.

**Comment by CyrusNajmabadi** [comment]
> @CyrusNajmabadi the issue is the examples and SyntaxEditorBasedCodeFix use the internal API's so to use them as a starting point I need to clone the internal API's. Also using SyntaxEditor as a starting point is not something that can be "whipped up rather simply" without internal knowledge

True.  I should have been clearer.  What is more accurate is: this can be whipped up simply, if one is already familiar with the public SyntaxEditor API.  

> Since I am trying to fix standard VB errors like ERR_StrictDisallowImplicitObject and it does not ignore Generated Code 

What is "it" in your "and it does not ignore" clause mean?  Thanks!

**Comment by sharwell** [comment]
@paul1956 As an alternative, consider looking at the fixes in StyleCop Analyzers. The [StyleCop Analyzers Status](https://stylecop.pdelvo.com/) page has a code fix column. Code fixes with ⌛️ use a deterministic variant of Roslyn's original batch fixer, but the ones with ✔️ use targeted high-speed fixers that do not rely on the batch fixer and can handle hundreds of thousands of fixes without problems. The approach varies quite a bit by the specific actions taken during the fix, but in most cases the common implementation strategy is to fix everything in one file as a single operation.

:memo: The status site appears to be down right now, but you can also look in the code for types derived from [`DocumentBasedFixAllProvider`](https://github.com/DotNetAnalyzers/StyleCopAnalyzers/blob/master/StyleCop.Analyzers/StyleCop.Analyzers.CodeFixes/Helpers/DocumentBasedFixAllProvider.cs).

**Comment by paul1956** [comment]
@CyrusNajmabadi  I have a code fix provider and am trying to fix error ERR_StrictDisallowImplicitObject for example, if I set a project to require As Clause on every declaration the compiler correctly reports them on generated code, this is especially true in Forms projects and Roslyn for example. But trying to fix those is a waste of time because they get regenerated. So I want my batch fixer to ignore Generated Files, but I don't see a way to do that without cloning one of the IsGenerateCode routines.

@sharwell I not have a CodeRefactoringProvider that can easily and quickly batch fix all the issues in a document using the DocumentEditor (I used RemoveByVal as a sample). I have also got a very fast CodeFix integrated using SyntaxEditorBasedCodeFix but it doesn't every return any changes (it does all the work but the UI shows no changes), I suspect I am not overriding something that needs to be overridden, and comes from the following code because fixAllContext.State is internal.
```
            public sealed override async Task<CodeAction> GetFixAsync(FixAllContext fixAllContext)
            {
                var documentsAndDiagnosticsToFixMap = await fixAllContext.GetDocumentDiagnosticsToFixAsync().ConfigureAwait(false);
                return await GetFixAsync(documentsAndDiagnosticsToFixMap, fixAllContext.State, fixAllContext.CancellationToken).ConfigureAwait(false);
            }

            internal sealed override async Task<CodeAction> GetFixAsync(
                ImmutableDictionary<Document, ImmutableArray<Diagnostic>> documentsAndDiagnosticsToFixMap,
                FixAllState fixAllState,
                CancellationToken cancellationToken)
            {
```
The actions are very simple replace one (mostly VariableDeclarationSyntax) node with another of the same type.

**Comment by sharwell** [comment]
@paul1956 Here's an example of a Fix All Provider which removes `this.` from locations where it is not required:

https://github.com/DotNetAnalyzers/StyleCopAnalyzers/blob/79abcc6a19554dc0148bc2afa14dfb97e1f9a02c/StyleCop.Analyzers/StyleCop.Analyzers.CodeFixes/ReadabilityRules/SX1101CodeFixProvider.cs#L68-L100

**Comment by paul1956** [comment]
@sharwell thanks for the example, before I try another implementation I would like to understand how the SyntaxEdit is supposed to work. I am using your DocumentBasedFixAllProvider as a starting point and when I look at Editor.GetChangedRoot it is always the original document.
```
       Protected Overrides Async Function FixAllInDocumentAsync(_FixAllContext As FixAllContext, _Document As Document, diagnostics As ImmutableArray(Of Diagnostic)) As Task(Of SyntaxNode)
            Dim Editor As SyntaxEditor = Await DocumentEditor.CreateAsync(_Document)
            Dim oldRoot As SyntaxNode = Editor.OriginalRoot
           '  More Setup
            For Each _diagnostic As Diagnostic In diagnostics
                  Dim oldNameNode As SyntaxNode = _diagnostic.Location.FindNode(getInnermostNodeForTie:=True, cancellationToken:=_CancellationToken)
                  Dim diagnosticSpanStart As Integer = _diagnostic.Location.SourceSpan.Start
                  Dim declarationSyntax As SyntaxNode = oldRoot.FindToken(diagnosticSpanStart).Parent.FirstAncestorOrSelfOfType(GetType(VariableDeclaratorSyntax))
                  ' Get replacement and replace
                   Dim NewVariableDeclarationNode As SyntaxNode = GetNewVariableDeclaratorSyntax()
                   '  NewVariableDeclarationNode  has the correct change
                   Editor.ReplaceNode(declarationSyntax, NewVariableDeclarationNode)
            Next _diagnostic
            Return Editor.GetChangedRoot
        End Function
 ```

**Comment by paul1956** [comment]
I am going to close this, it looks like between the 2.4 fixes and the examples by @sharwell most people should be able to implement something that doesn't run out of memory. Plus #22591 should address the missing testing for batch fixers over large projects. I would still like to understand why SyntaxEdit doesn't work as expected but that is really a separate issue.

**Comment by CyrusNajmabadi** [comment]
> I would still like to understand why SyntaxEdit doesn't work as expected but that is really a separate issue.

If you debug into your code, do you see the calls to .ReplaceNode happening?

**Comment by paul1956** [comment]
@CyrusNajmabadi  Yes and I debug.Write the old and new nodes and verify they are always the same type (mostly NewVariableDeclaratorSyntax) differing only by presence or absence of  an AsClause. The extracted code about has no If statements before the Editor.ReplaceNode and no Try Block so it would throw.

**Comment by CyrusNajmabadi** [comment]
That is very strange then.  I think a repro project would have to be made in order for someone to dive in an see what was wrong.  Alternatively, if you wanted to debug into SyntaxEditor itself, that might reveal why no change is occurring.

**Comment by paul1956** [comment]
@CyrusNajmabadi I have been asking since Roslyn 1.0 how to debug my code into Roslyn and have yet to see anything for someone outside the development team to be able to do it. Since 2.0 I can finally build Roslyn from Master without errors because of excellent easy to follow instructions and provided scripts.

The good news is that with the solution above from @sharwell  I can do a BatchFixer over Roslyn Samples and make over 11,000 changes in under a minute without any memory issues. Using the WellKnownFixAllProviders.BatchFixer provider I can't do a few 100 in many hours without running out of memory (though 2.4 is much better).

**Comment by CyrusNajmabadi** [comment]
> I have been asking since Roslyn 1.0 how to debug my code into Roslyn

I just get the roslyn sources, add my code to it, choose "ServicesVisualStudio" as the startup project, build everything, and hit f5.  Does that not work for you?

**Comment by CyrusNajmabadi** [comment]
> The good news is that with the solution above from @sharwell I can do a BatchFixer over Roslyn Samples and make over 11,000 changes in under a minute without any memory issues. 

Great!  Looking at that code, it's the same sort of approach that we use with the SyntaxEditor solution.  Both simply build up a set of work to do and just apply them to the document in one go.  The only real benefit of the SyntaxEditor approach is that it works well for fixes that need to go beyond just doing ReplaceNodes.  i.e. some fixes involve multiple add/remove/insert/replace operations.  If you don't need that, then no need to pull in SyntaxEditor.

> WellKnownFixAllProviders.BatchFixer provider I can't do a few 100 in many hours without running out of memory (though 2.4 is much better).

Yup.  The design of fixers makes a generalized 'batch' solution inherently hard to make fast.  First, because each fix is only valid against original snapshot, you have to perform N fixes for N diagnostics, unlike teh approaches Sam and I showed (where you can do 1 fix for N diagnostics).  Then, because all we have from those fixes are essentially text changes, it's necessary to try to merge those together (something which is not always possible).  

Because of these general inefficiencies, it's only recommended to use the built in batch fixer if fix-all is generally only going to fix a handful of cases per file max.  If your fixer would make many changes per file, it is definitely recommended to just write a fixer yourself that processes all your diagnostics once against a single tree.  

On top of what i wrote above, this approach may also be necessary for correctness.  For example, say this is hte "Remove unnecessary cast" fix.  If the user has code like "(double)d == (double)e", then there may be two diagnostics reported (stating that each cast *individually* is unnecessary).  However, if you then go and remove both, it may change the meaning.  As such, the batch fixer will want to go correct all of these in bulk, but also ensure that it only fixes things up if it is still safe to do so given the previous work it did.  

This sort of thing is complex, is domain specific to the fix, and can have a big impact on perf.  That's why it's hard to provide generalized solutions here.

**Comment by paul1956** [comment]
@CyrusNajmabadi built Roslyn, did Add Existing for my project, built Roslyn again, set ServiceVisualStudio as Startup Project and I get "A Project with an Output Type of Class cannot be started directly, I think I am missing something. Also do I need to change anything in my project to get the Roslyn Dll's from the Master I built instead of NuGet? Most of my Reference Point to NuGet packages.

**Comment by sharwell** [comment]
@paul1956 I believe the startup project should be VisualStudioSetup.Next

**Comment by paul1956** [comment]
@sharwell this does not answer the other issue, how do I find all the Roslyn DLL's that I was getting from NuGet that I now need to get from Roslyn Master source?
